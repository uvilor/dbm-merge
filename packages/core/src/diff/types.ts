import type { Check, Column, ForeignKey, Index, Routine, Table, Trigger, View } from '../types';

export interface ChangeEnvelope<TChange> {
  name: string;
  change: TChange;
}

export interface ColumnChange {
  typeChanged?: { from: string; to: string };
  nullableChanged?: { from: boolean; to: boolean };
  defaultChanged?: { from: string | null | undefined; to: string | null | undefined };
  generatedChanged?: { from: Column['generated']; to: Column['generated'] };
  collationChanged?: { from: string | null | undefined; to: string | null | undefined };
}

export interface PrimaryKeyChange {
  from?: Table['primaryKey'];
  to?: Table['primaryKey'];
}

export interface IndexChange {
  from: Index;
  to: Index;
}

export interface ForeignKeyChange {
  from: ForeignKey;
  to: ForeignKey;
}

export interface CheckChange {
  from: Check;
  to: Check;
}

export interface DefinitionChange {
  from: string;
  to: string;
}

export interface TableChange {
  table: Table;
  sourceTable: Table;
  addedColumns: Column[];
  removedColumns: Column[];
  columnChanges: ChangeEnvelope<ColumnChange>[];
  primaryKeyChange?: PrimaryKeyChange;
  addedIndexes: Index[];
  removedIndexes: Index[];
  indexChanges: ChangeEnvelope<IndexChange>[];
  addedChecks: Check[];
  removedChecks: Check[];
  checkChanges: ChangeEnvelope<CheckChange>[];
  addedForeignKeys: ForeignKey[];
  removedForeignKeys: ForeignKey[];
  foreignKeyChanges: ChangeEnvelope<ForeignKeyChange>[];
}

export interface ObjectCollectionDiff<T, TChange = DefinitionChange> {
  added: T[];
  removed: T[];
  changed: ChangeEnvelope<TChange>[];
}

export interface DiffResult {
  tables: {
    added: Table[];
    removed: Table[];
    changed: TableChange[];
  };
  views: ObjectCollectionDiff<View>;
  routines: ObjectCollectionDiff<Routine>;
  triggers: ObjectCollectionDiff<Trigger>;
}
