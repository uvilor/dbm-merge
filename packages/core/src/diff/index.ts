import isEqual from 'lodash/isEqual.js';
import sortBy from 'lodash/sortBy.js';

import type { SchemaModel, Table } from '../types';
import type {
  ChangeEnvelope,
  ColumnChange,
  DefinitionChange,
  DiffResult,
  TableChange,
} from './types';
import {
  compareChecks,
  compareColumns,
  compareForeignKeys,
  compareIndexes,
  comparePrimaryKeys,
} from './tableComparators';

export { type DiffResult } from './types';

export function computeDiff(a: SchemaModel, b: SchemaModel): DiffResult {
  const tableDiff = diffTables(a.tables, b.tables);
  const viewDiff = diffCollection(a.views, b.views, (item) => item.name, (item) => item.definition);
  const routineDiff = diffCollection(
    a.routines,
    b.routines,
    (item) => `${item.kind}:${item.name}`,
    (item) => item.definition,
  );
  const triggerDiff = diffCollection(
    a.triggers,
    b.triggers,
    (item) => `${item.table}.${item.name}`,
    (item) => [item.timing, item.events.join(','), item.definition].join('\n'),
  );

  return {
    tables: tableDiff,
    views: viewDiff,
    routines: routineDiff,
    triggers: triggerDiff,
  };
}

function diffTables(source: Table[], target: Table[]) {
  const sourceMap = new Map(source.map((table) => [table.name, table]));
  const targetMap = new Map(target.map((table) => [table.name, table]));

  const added: Table[] = [];
  const removed: Table[] = [];
  const changed: TableChange[] = [];

  target.forEach((table) => {
    if (!sourceMap.has(table.name)) {
      added.push(table);
    }
  });

  source.forEach((table) => {
    if (!targetMap.has(table.name)) {
      removed.push(table);
    }
  });

  source.forEach((table) => {
    const counterpart = targetMap.get(table.name);
    if (!counterpart) return;
    const tableChange = diffTable(table, counterpart);
    if (tableChange) {
      changed.push(tableChange);
    }
  });

  return { added, removed, changed: sortBy(changed, (change) => change.table.name) };
}

function diffTable(source: Table, target: Table): TableChange | null {
  const { added: addedColumns, removed: removedColumns, changed: columnChanges } = compareColumns(
    source,
    target,
  );

  const primaryKeyChange = comparePrimaryKeys(source.primaryKey, target.primaryKey);
  const { added: addedIndexes, removed: removedIndexes, changed: indexChanges } = compareIndexes(
    source.indexes,
    target.indexes,
  );
  const { added: addedChecks, removed: removedChecks, changed: checkChanges } = compareChecks(
    source.checks,
    target.checks,
  );
  const {
    added: addedForeignKeys,
    removed: removedForeignKeys,
    changed: foreignKeyChanges,
  } = compareForeignKeys(source.fks, target.fks);

  const hasChanges =
    addedColumns.length > 0 ||
    removedColumns.length > 0 ||
    columnChanges.length > 0 ||
    primaryKeyChange != null ||
    addedIndexes.length > 0 ||
    removedIndexes.length > 0 ||
    indexChanges.length > 0 ||
    addedChecks.length > 0 ||
    removedChecks.length > 0 ||
    checkChanges.length > 0 ||
    addedForeignKeys.length > 0 ||
    removedForeignKeys.length > 0 ||
    foreignKeyChanges.length > 0;

  if (!hasChanges) {
    return null;
  }

  return {
    table: target,
    sourceTable: source,
    addedColumns,
    removedColumns,
    columnChanges,
    primaryKeyChange: primaryKeyChange ?? undefined,
    addedIndexes,
    removedIndexes,
    indexChanges,
    addedChecks,
    removedChecks,
    checkChanges,
    addedForeignKeys,
    removedForeignKeys,
    foreignKeyChanges,
  };
}

function diffCollection<T>(
  source: T[],
  target: T[],
  keyFn: (item: T) => string,
  valueFn: (item: T) => unknown,
): { added: T[]; removed: T[]; changed: ChangeEnvelope<DefinitionChange>[] } {
  const sourceMap = new Map<string, T>();
  const targetMap = new Map<string, T>();

  source.forEach((item) => {
    sourceMap.set(keyFn(item), item);
  });

  target.forEach((item) => {
    targetMap.set(keyFn(item), item);
  });

  const added: T[] = [];
  const removed: T[] = [];
  const changed: ChangeEnvelope<DefinitionChange>[] = [];

  targetMap.forEach((value, key) => {
    if (!sourceMap.has(key)) {
      added.push(value);
    }
  });

  sourceMap.forEach((value, key) => {
    if (!targetMap.has(key)) {
      removed.push(value);
    }
  });

  sourceMap.forEach((value, key) => {
    if (!targetMap.has(key)) return;
    const targetValue = targetMap.get(key)!;
    if (!isEqual(valueFn(value), valueFn(targetValue))) {
      changed.push({
        name: key,
        change: {
          from: String(valueFn(value)),
          to: String(valueFn(targetValue)),
        },
      });
    }
  });

  return {
    added: sortBy(added, (item) => keyFn(item)),
    removed: sortBy(removed, (item) => keyFn(item)),
    changed: sortBy(changed, (item) => item.name),
  };
}
