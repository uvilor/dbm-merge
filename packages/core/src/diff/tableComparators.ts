import isEqual from 'lodash/isEqual.js';
import sortBy from 'lodash/sortBy.js';

import type { Check, Column, ForeignKey, Index, Table } from '../types';
import type {
  ChangeEnvelope,
  CheckChange,
  ColumnChange,
  ForeignKeyChange,
  IndexChange,
  PrimaryKeyChange,
} from './types';

export function compareColumns(
  source: Table,
  target: Table,
): {
  added: Column[];
  removed: Column[];
  changed: ChangeEnvelope<ColumnChange>[];
} {
  const sourceMap = new Map(source.columns.map((column) => [column.name, column]));
  const targetMap = new Map(target.columns.map((column) => [column.name, column]));

  const added: Column[] = [];
  const removed: Column[] = [];
  const changed: ChangeEnvelope<ColumnChange>[] = [];

  target.columns.forEach((column) => {
    if (!sourceMap.has(column.name)) {
      added.push(column);
    }
  });

  source.columns.forEach((column) => {
    if (!targetMap.has(column.name)) {
      removed.push(column);
    }
  });

  source.columns.forEach((column) => {
    const counterpart = targetMap.get(column.name);
    if (!counterpart) return;
    const change = diffColumn(column, counterpart);
    if (change) {
      changed.push({
        name: column.name,
        change,
      });
    }
  });

  return {
    added: sortBy(added, (col) => col.name),
    removed: sortBy(removed, (col) => col.name),
    changed: sortBy(changed, (change) => change.name),
  };
}

function diffColumn(source: Column, target: Column): ColumnChange | null {
  const change: ColumnChange = {};
  if (source.dataType !== target.dataType) {
    change.typeChanged = { from: source.dataType, to: target.dataType };
  }
  if (source.nullable !== target.nullable) {
    change.nullableChanged = { from: source.nullable, to: target.nullable };
  }
  if (!isEqual(source.default ?? null, target.default ?? null)) {
    change.defaultChanged = { from: source.default, to: target.default };
  }
  if ((source.generated ?? null) !== (target.generated ?? null)) {
    change.generatedChanged = { from: source.generated ?? null, to: target.generated ?? null };
  }
  if ((source.collation ?? null) !== (target.collation ?? null)) {
    change.collationChanged = { from: source.collation ?? null, to: target.collation ?? null };
  }

  return Object.keys(change).length > 0 ? change : null;
}

export function comparePrimaryKeys(
  source?: Table['primaryKey'],
  target?: Table['primaryKey'],
): PrimaryKeyChange | null {
  if (!source && !target) return null;
  if (!source && target) return { to: target };
  if (source && !target) return { from: source };
  if (!isEqual(sortBy(source!.columns), sortBy(target!.columns))) {
    return {
      from: source!,
      to: target!,
    };
  }
  return null;
}

export function compareIndexes(
  source: Index[],
  target: Index[],
): {
  added: Index[];
  removed: Index[];
  changed: ChangeEnvelope<IndexChange>[];
} {
  const sourceMap = new Map(source.map((index) => [index.name, index]));
  const targetMap = new Map(target.map((index) => [index.name, index]));

  const added: Index[] = [];
  const removed: Index[] = [];
  const changed: ChangeEnvelope<IndexChange>[] = [];

  target.forEach((index) => {
    if (!sourceMap.has(index.name)) {
      added.push(index);
    }
  });

  source.forEach((index) => {
    if (!targetMap.has(index.name)) {
      removed.push(index);
    }
  });

  source.forEach((index) => {
    const targetIndex = targetMap.get(index.name);
    if (!targetIndex) return;
    if (!isEqual(normalizeIndex(index), normalizeIndex(targetIndex))) {
      changed.push({ name: index.name, change: { from: index, to: targetIndex } });
    }
  });

  return {
    added: sortBy(added, (index) => index.name),
    removed: sortBy(removed, (index) => index.name),
    changed: sortBy(changed, (index) => index.name),
  };
}

function normalizeIndex(index: Index) {
  return {
    unique: index.unique,
    using: index.using?.toLowerCase() ?? null,
    columns: sortBy(index.columns.map((col) => col.toLowerCase())),
  };
}

export function compareChecks(
  source: Check[],
  target: Check[],
): {
  added: Check[];
  removed: Check[];
  changed: ChangeEnvelope<CheckChange>[];
} {
  const sourceMap = new Map(source.map((check) => [check.name, check]));
  const targetMap = new Map(target.map((check) => [check.name, check]));

  const added: Check[] = [];
  const removed: Check[] = [];
  const changed: ChangeEnvelope<CheckChange>[] = [];

  target.forEach((check) => {
    if (!sourceMap.has(check.name)) {
      added.push(check);
    }
  });

  source.forEach((check) => {
    if (!targetMap.has(check.name)) {
      removed.push(check);
    }
  });

  source.forEach((check) => {
    const targetCheck = targetMap.get(check.name);
    if (!targetCheck) return;
    if (!isEqual(normalizeExpression(check.expression), normalizeExpression(targetCheck.expression))) {
      changed.push({ name: check.name, change: { from: check, to: targetCheck } });
    }
  });

  return {
    added: sortBy(added, (check) => check.name),
    removed: sortBy(removed, (check) => check.name),
    changed: sortBy(changed, (check) => check.name),
  };
}

function normalizeExpression(expression: string) {
  return expression.replace(/\s+/g, ' ').trim();
}

export function compareForeignKeys(
  source: ForeignKey[],
  target: ForeignKey[],
): {
  added: ForeignKey[];
  removed: ForeignKey[];
  changed: ChangeEnvelope<ForeignKeyChange>[];
} {
  const sourceMap = new Map(source.map((fk) => [fk.name, fk]));
  const targetMap = new Map(target.map((fk) => [fk.name, fk]));

  const added: ForeignKey[] = [];
  const removed: ForeignKey[] = [];
  const changed: ChangeEnvelope<ForeignKeyChange>[] = [];

  target.forEach((fk) => {
    if (!sourceMap.has(fk.name)) {
      added.push(fk);
    }
  });

  source.forEach((fk) => {
    if (!targetMap.has(fk.name)) {
      removed.push(fk);
    }
  });

  source.forEach((fk) => {
    const targetFk = targetMap.get(fk.name);
    if (!targetFk) return;
    if (!isEqual(normalizeForeignKey(fk), normalizeForeignKey(targetFk))) {
      changed.push({ name: fk.name, change: { from: fk, to: targetFk } });
    }
  });

  return {
    added: sortBy(added, (fk) => fk.name),
    removed: sortBy(removed, (fk) => fk.name),
    changed: sortBy(changed, (fk) => fk.name),
  };
}

function normalizeForeignKey(fk: ForeignKey) {
  return {
    columns: sortBy(fk.columns.map((col) => col.toLowerCase())),
    refTable: fk.refTable.toLowerCase(),
    refColumns: sortBy(fk.refColumns.map((col) => col.toLowerCase())),
    onUpdate: fk.onUpdate?.toUpperCase() ?? null,
    onDelete: fk.onDelete?.toUpperCase() ?? null,
  };
}
