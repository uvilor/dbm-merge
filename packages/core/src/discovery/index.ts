export { loadPostgres } from './postgres';
export { loadMariaDB } from './mariadb';
