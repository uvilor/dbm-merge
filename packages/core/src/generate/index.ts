import { EOL } from 'node:os';

import type { DiffResult } from '../diff';
import type { Column, Index, Table } from '../types';

export interface GenOptions {
  direction: 'AtoB' | 'BtoA';
  withTransaction?: boolean;
  safeMode?: boolean;
  ifExists?: boolean;
  cascade?: boolean;
}

interface TableChangePlan {
  table: Table;
  desired: Table;
  addColumns: Column[];
  dropColumns: Column[];
  alterColumns: Array<{ column: Column; change: string[] }>;
  dropIndexes: Index[];
  createIndexes: Index[];
}

export function toPostgres(diff: DiffResult, opts: GenOptions): string {
  const lines: string[] = [];
  const { toCreate, toDrop, tablePlans } = planTables(diff, opts);

  if (opts.withTransaction) {
    lines.push('BEGIN;');
  }

  if (opts.safeMode && toDrop.length > 0) {
    lines.push('-- SAFE MODE: review drops below before executing.');
  }

  toDrop.forEach((table) => {
    const drop = `DROP TABLE${opts.ifExists ? ' IF EXISTS' : ''} ${quoteIdent(table.name)}${
      opts.cascade ? ' CASCADE' : ''
    };`;
    lines.push(opts.safeMode ? `-- ${drop}` : drop);
  });

  toCreate.forEach((table) => {
    lines.push(renderPostgresCreateTable(table));
  });

  tablePlans.forEach((plan) => {
    const tableName = quoteIdent(plan.table.name);
    plan.dropColumns.forEach((column) => {
      const drop = `ALTER TABLE ${tableName} DROP COLUMN ${quoteIdent(column.name)}${opts.cascade ? ' CASCADE' : ''};`;
      lines.push(opts.safeMode ? `-- ${drop}` : drop);
    });
    plan.addColumns.forEach((column) => {
      lines.push(`ALTER TABLE ${tableName} ADD COLUMN ${renderPostgresColumn(column)};`);
    });
    plan.alterColumns.forEach((alteration) => {
      alteration.change.forEach((ddl) => {
        lines.push(`ALTER TABLE ${tableName} ${ddl};`);
      });
    });

    plan.dropIndexes.forEach((index) => {
      const drop = `DROP INDEX${opts.ifExists ? ' IF EXISTS' : ''} ${quoteIdent(index.name)};`;
      lines.push(opts.safeMode ? `-- ${drop}` : drop);
    });
    plan.createIndexes.forEach((index) => {
      lines.push(renderPostgresIndex(plan.table.name, index));
    });
  });

  renderViewChanges(diff, opts, lines, 'postgres');
  renderRoutineChanges(diff, opts, lines, 'postgres');
  renderTriggerChanges(diff, opts, lines, 'postgres');

  if (opts.withTransaction) {
    lines.push('COMMIT;');
  }

  return lines.filter((line) => line.trim().length > 0).join(EOL + EOL);
}

export function toMariaDB(diff: DiffResult, opts: GenOptions): string {
  const lines: string[] = [];
  const { toCreate, toDrop, tablePlans } = planTables(diff, opts);

  if (opts.withTransaction) {
    lines.push('START TRANSACTION;');
  }

  if (opts.safeMode && toDrop.length > 0) {
    lines.push('-- SAFE MODE: review drops below before executing.');
  }

  toDrop.forEach((table) => {
    const drop = `DROP TABLE${opts.ifExists ? ' IF EXISTS' : ''} ${backtick(table.name)}${
      opts.cascade ? ' CASCADE' : ''
    };`;
    lines.push(opts.safeMode ? `-- ${drop}` : drop);
  });

  toCreate.forEach((table) => {
    lines.push(renderMariaCreateTable(table));
  });

  tablePlans.forEach((plan) => {
    const tableName = backtick(plan.table.name);
    plan.dropColumns.forEach((column) => {
      const drop = `ALTER TABLE ${tableName} DROP COLUMN ${backtick(column.name)};`;
      lines.push(opts.safeMode ? `-- ${drop}` : drop);
    });
    plan.addColumns.forEach((column) => {
      lines.push(`ALTER TABLE ${tableName} ADD COLUMN ${renderMariaColumn(column)};`);
    });
    plan.alterColumns.forEach((alteration) => {
      alteration.change.forEach((ddl) => {
        lines.push(`ALTER TABLE ${tableName} ${ddl};`);
      });
    });

    plan.dropIndexes.forEach((index) => {
      const drop = `DROP INDEX ${backtick(index.name)} ON ${tableName};`;
      lines.push(opts.safeMode ? `-- ${drop}` : drop);
    });
    plan.createIndexes.forEach((index) => {
      lines.push(renderMariaIndex(plan.table.name, index));
    });
  });

  renderViewChanges(diff, opts, lines, 'mariadb');
  renderRoutineChanges(diff, opts, lines, 'mariadb');
  renderTriggerChanges(diff, opts, lines, 'mariadb');

  if (opts.withTransaction) {
    lines.push('COMMIT;');
  }

  return lines.filter((line) => line.trim().length > 0).join(EOL + EOL);
}

function planTables(diff: DiffResult, opts: GenOptions) {
  const createTables = opts.direction === 'AtoB' ? diff.tables.removed : diff.tables.added;
  const dropTables = opts.direction === 'AtoB' ? diff.tables.added : diff.tables.removed;

  const tablePlans = diff.tables.changed.map((change) => buildTablePlan(change, opts.direction));
  return { toCreate: createTables, toDrop: dropTables, tablePlans };
}

function buildTablePlan(
  change: DiffResult['tables']['changed'][number],
  direction: GenOptions['direction'],
): TableChangePlan {
  const desired = direction === 'AtoB' ? change.sourceTable : change.table;
  const working = direction === 'AtoB' ? change.table : change.sourceTable;

  const addColumns = direction === 'AtoB' ? change.removedColumns : change.addedColumns;
  const dropColumns = direction === 'AtoB' ? change.addedColumns : change.removedColumns;

  const alterColumns = change.columnChanges.map(({ name, change: columnChange }) => {
    const desiredColumn = desired.columns.find((col) => col.name === name) ?? desired.columns[0];
    if (!desiredColumn) {
      return { column: working.columns.find((col) => col.name === name)!, change: [`-- TODO: review column ${name}`] };
    }
    const statements: string[] = [];
    if (columnChange.typeChanged) {
      statements.push(`ALTER COLUMN ${quoteIdent(name)} TYPE ${desiredColumn.dataType}`);
      statements.push(`-- TODO: verify casts for ${name}`);
    }
    if (columnChange.nullableChanged) {
      const clause = desiredColumn.nullable ? 'DROP NOT NULL' : 'SET NOT NULL';
      statements.push(`ALTER COLUMN ${quoteIdent(name)} ${clause}`);
    }
    if (columnChange.defaultChanged) {
      const def = desiredColumn.default;
      const clause = def == null ? 'DROP DEFAULT' : `SET DEFAULT ${def}`;
      statements.push(`ALTER COLUMN ${quoteIdent(name)} ${clause}`);
    }
    if (columnChange.generatedChanged) {
      statements.push(`-- TODO: reconcile generation strategy for ${name}`);
    }
    if (columnChange.collationChanged) {
      statements.push(`-- TODO: adjust collation for ${name}`);
    }
    return { column: desiredColumn, change: statements };
  });

  const dropIndexes = direction === 'AtoB' ? change.addedIndexes : change.removedIndexes;
  const createIndexes = direction === 'AtoB' ? change.removedIndexes : change.addedIndexes;

  return {
    table: working,
    desired,
    addColumns,
    dropColumns,
    alterColumns,
    dropIndexes,
    createIndexes,
  };
}

function renderPostgresCreateTable(table: Table): string {
  const columns = table.columns.map((column) => renderPostgresColumn(column));
  if (table.primaryKey) {
    columns.push(
      `CONSTRAINT ${quoteIdent(table.primaryKey.name ?? `${table.name}_pkey`)} PRIMARY KEY (${table.primaryKey.columns
        .map(quoteIdent)
        .join(', ')})`,
    );
  }
  return `CREATE TABLE ${quoteIdent(table.name)} (\n  ${columns.join(',\n  ')}\n);`;
}

function renderMariaCreateTable(table: Table): string {
  const columns = table.columns.map((column) => renderMariaColumn(column));
  if (table.primaryKey) {
    columns.push(
      `CONSTRAINT ${backtick(table.primaryKey.name ?? `${table.name}_pk`)} PRIMARY KEY (${table.primaryKey.columns
        .map((col) => backtick(col))
        .join(', ')})`,
    );
  }
  return `CREATE TABLE ${backtick(table.name)} (\n  ${columns.join(',\n  ')}\n) ENGINE=InnoDB;`;
}

function renderPostgresColumn(column: Column): string {
  const parts = [quoteIdent(column.name), column.dataType];
  if (column.length) {
    parts[1] = `${column.dataType}(${column.length})`;
  }
  if (column.precisionScale) {
    parts[1] = `${column.dataType}(${column.precisionScale.precision}, ${column.precisionScale.scale})`;
  }
  if (!column.nullable) {
    parts.push('NOT NULL');
  }
  if (column.default != null) {
    parts.push(`DEFAULT ${column.default}`);
  }
  if (column.generated === 'identity' || column.generated === 'sequence') {
    parts.push('-- TODO: ensure generation strategy is preserved');
  }
  if (column.collation) {
    parts.push(`COLLATE ${quoteIdent(column.collation)}`);
  }
  return parts.join(' ');
}

function renderMariaColumn(column: Column): string {
  const parts = [backtick(column.name), column.dataType.toUpperCase()];
  if (column.length) {
    parts[1] = `${column.dataType.toUpperCase()}(${column.length})`;
  }
  if (column.precisionScale) {
    parts[1] = `${column.dataType.toUpperCase()}(${column.precisionScale.precision}, ${column.precisionScale.scale})`;
  }
  if (!column.nullable) {
    parts.push('NOT NULL');
  }
  if (column.default != null) {
    parts.push(`DEFAULT ${column.default}`);
  }
  if (column.generated === 'auto_increment') {
    parts.push('AUTO_INCREMENT');
  }
  if (column.collation) {
    parts.push(`COLLATE ${column.collation}`);
  }
  return parts.join(' ');
}

function renderPostgresIndex(tableName: string, index: Index): string {
  return `CREATE${index.unique ? ' UNIQUE' : ''} INDEX ${quoteIdent(index.name)} ON ${quoteIdent(tableName)}${
    index.using ? ` USING ${index.using}` : ''
  } (${index.columns.map(quoteIdent).join(', ')});`;
}

function renderMariaIndex(tableName: string, index: Index): string {
  const type = index.unique ? 'UNIQUE ' : '';
  return `CREATE ${type}INDEX ${backtick(index.name)} ON ${backtick(tableName)} (${index.columns
    .map((column) => backtick(column))
    .join(', ')});`;
}

function renderViewChanges(
  diff: DiffResult,
  opts: GenOptions,
  lines: string[],
  dialect: 'postgres' | 'mariadb',
) {
  const createViews = opts.direction === 'AtoB' ? diff.views.removed : diff.views.added;
  const dropViews = opts.direction === 'AtoB' ? diff.views.added : diff.views.removed;

  dropViews.forEach((view) => {
    const drop = `DROP VIEW${opts.ifExists ? ' IF EXISTS' : ''} ${
      dialect === 'postgres' ? quoteIdent(view.name) : backtick(view.name)
    }${opts.cascade ? ' CASCADE' : ''};`;
    lines.push(opts.safeMode ? `-- ${drop}` : drop);
  });

  createViews.forEach((view) => {
    const name = dialect === 'postgres' ? quoteIdent(view.name) : backtick(view.name);
    lines.push(`CREATE OR REPLACE VIEW ${name} AS\n${view.definition.trim()};`);
  });

  diff.views.changed.forEach((change) => {
    lines.push(`-- TODO: view ${change.name} changed; consider drop and recreate.`);
  });
}

function renderRoutineChanges(
  diff: DiffResult,
  opts: GenOptions,
  lines: string[],
  dialect: 'postgres' | 'mariadb',
) {
  const createRoutines = opts.direction === 'AtoB' ? diff.routines.removed : diff.routines.added;
  const dropRoutines = opts.direction === 'AtoB' ? diff.routines.added : diff.routines.removed;

  dropRoutines.forEach((routine) => {
    const drop = `DROP ${routine.kind.toUpperCase()}${opts.ifExists ? ' IF EXISTS' : ''} ${
      dialect === 'postgres' ? quoteIdent(routine.name) : backtick(routine.name)
    };`;
    lines.push(opts.safeMode ? `-- ${drop}` : drop);
  });

  createRoutines.forEach((routine) => {
    lines.push(`-- Routine ${routine.name}`);
    lines.push(routine.definition.trim());
  });

  diff.routines.changed.forEach((change) => {
    lines.push(`-- TODO: routine ${change.name} definition changed; drop and recreate manually.`);
  });
}

function renderTriggerChanges(
  diff: DiffResult,
  opts: GenOptions,
  lines: string[],
  dialect: 'postgres' | 'mariadb',
) {
  const createTriggers = opts.direction === 'AtoB' ? diff.triggers.removed : diff.triggers.added;
  const dropTriggers = opts.direction === 'AtoB' ? diff.triggers.added : diff.triggers.removed;

  dropTriggers.forEach((trigger) => {
    const name = dialect === 'postgres' ? quoteIdent(trigger.name) : backtick(trigger.name);
    const table = dialect === 'postgres' ? quoteIdent(trigger.table) : backtick(trigger.table);
    const drop = `DROP TRIGGER${opts.ifExists ? ' IF EXISTS' : ''} ${name} ON ${table};`;
    lines.push(opts.safeMode ? `-- ${drop}` : drop);
  });

  createTriggers.forEach((trigger) => {
    lines.push(`-- Trigger ${trigger.name} on ${trigger.table}`);
    lines.push(trigger.definition.trim());
  });

  diff.triggers.changed.forEach((change) => {
    lines.push(`-- TODO: trigger ${change.name} changed; consider drop and recreate.`);
  });
}

function quoteIdent(identifier: string): string {
  return `"${identifier.replace(/"/g, '""')}"`;
}

function backtick(identifier: string): string {
  return `\`${identifier.replace(/`/g, '``')}\``;
}
