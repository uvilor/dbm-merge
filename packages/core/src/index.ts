export type DatabaseKind = 'postgres' | 'mariadb';

export interface Column {
  name: string;
  dataType: string;
  nullable: boolean;
  default?: string | null;
}

export interface Table {
  name: string;
  columns: Column[];
}

export interface SchemaModel {
  tables: Table[];
}

export async function loadSchema(_conn: string, _engine: DatabaseKind): Promise<SchemaModel> {
  // Stub: en el MVP real, aquí se consulta information_schema/pg_catalog
  return { tables: [] };
}

export function computeDiff(a: SchemaModel, b: SchemaModel) {
  // Stub mínimo
  return {
    addedTables: b.tables.filter(tb => !a.tables.some(x => x.name === tb.name)).map(t => t.name),
    removedTables: a.tables.filter(tb => !b.tables.some(x => x.name === tb.name)).map(t => t.name)
  };
}

export function generateSql(_diff: ReturnType<typeof computeDiff>, _engine: DatabaseKind): string {
  // Stub mínimo de generador
  return `-- SQL de homologación (MVP stub)`;
}
