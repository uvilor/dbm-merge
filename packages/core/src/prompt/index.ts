import type { DatabaseKind } from '../types';

interface PromptInput {
  sourceKind: DatabaseKind;
  targetKind: DatabaseKind;
  diffSnippet: string;
  ddlProposal: string;
}

export function buildReviewPrompt({
  sourceKind,
  targetKind,
  diffSnippet,
  ddlProposal,
}: PromptInput): string {
  const safety = `You are assisting with a schema synchronization between ${sourceKind} and ${targetKind}.
Do not execute any SQL. Review the proposal for destructive operations and ensure they are explicitly confirmed.
Highlight risky statements (DROP, TRUNCATE, CASCADE) and suggest mitigations.`;

  return [
    '### Context',
    safety,
    '',
    '### Diff Summary',
    `\u0060\u0060\u0060\n${diffSnippet.trim()}\n\u0060\u0060\u0060`,
    '',
    '### Proposed DDL',
    `\u0060\u0060\u0060\n${ddlProposal.trim()}\n\u0060\u0060\u0060`,
    '',
    '### Tasks',
    '1. Validate that the DDL migrates the target schema to the source schema without data loss unless explicitly requested.',
    '2. Recommend improvements such as transactional wrappers, lock minimization, or phased rollouts.',
    '3. Identify ambiguous steps and propose safe execution notes or TODOs.',
    '4. Return the answer as a Markdown checklist summarizing the review and any outstanding items.',
  ].join('\n');
}

export type { PromptInput };
