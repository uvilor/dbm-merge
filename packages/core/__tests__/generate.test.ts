import { describe, expect, it } from 'vitest';

import { computeDiff } from '../src/diff';
import { toMariaDB, toPostgres } from '../src/generate';
import type { SchemaModel } from '../src/types';

const source: SchemaModel = {
  tables: [
    {
      name: 'users',
      columns: [
        { name: 'id', dataType: 'serial', nullable: false, generated: 'sequence' },
        { name: 'email', dataType: 'varchar', length: 255, nullable: false },
      ],
      primaryKey: { name: 'users_pkey', columns: ['id'] },
      indexes: [{ name: 'users_email_key', unique: true, columns: ['email'] }],
      checks: [],
      fks: [],
    },
  ],
  views: [],
  routines: [],
  triggers: [],
};

const target: SchemaModel = {
  tables: [
    {
      name: 'users',
      columns: [
        { name: 'id', dataType: 'serial', nullable: false, generated: 'sequence' },
        { name: 'email', dataType: 'varchar', length: 128, nullable: false },
        { name: 'status', dataType: 'varchar', length: 32, nullable: true, default: "'pending'" },
      ],
      primaryKey: { name: 'users_pkey', columns: ['id'] },
      indexes: [
        { name: 'users_email_key', unique: false, columns: ['email'] },
        { name: 'users_status_idx', unique: false, columns: ['status'] },
      ],
      checks: [],
      fks: [],
    },
  ],
  views: [],
  routines: [],
  triggers: [],
};

describe('DDL generators', () => {
  it('produces postgres DDL snapshots', () => {
    const diff = computeDiff(source, target);
    const sql = toPostgres(diff, { direction: 'AtoB', withTransaction: true, safeMode: true });
    expect(sql).toMatchSnapshot();
  });

  it('produces mariadb DDL snapshots', () => {
    const diff = computeDiff(source, target);
    const sql = toMariaDB(diff, { direction: 'AtoB', withTransaction: true, safeMode: false });
    expect(sql).toMatchSnapshot();
  });
});
