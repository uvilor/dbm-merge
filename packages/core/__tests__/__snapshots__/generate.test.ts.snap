// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DDL generators > produces mariadb DDL snapshots 1`] = `
"START TRANSACTION;

ALTER TABLE \`users\` DROP COLUMN \`status\`;

DROP INDEX \`users_status_idx\` ON \`users\`;

COMMIT;"
`;

exports[`DDL generators > produces postgres DDL snapshots 1`] = `
"BEGIN;

-- ALTER TABLE "users" DROP COLUMN "status";

-- DROP INDEX "users_status_idx";

COMMIT;"
`;
